### DISCUSSION TITLE
how to properly run a program and supply command line arguments?

### DISCUSSION THREAD

**Discussion by mortax** [winglang/wing#4939] [General]
i have a tiny program but the only way i am able to _run_ it is using `wing compile`, e.g.:

```
[univac]$ cat echo.main.w 
let main = (args: Array<str>) => {
  log(args.at(0));
  log(args.at(1));
};

main(["first", "two"]);
[univac]$ wing compile echo.main.w --platform sim           
first
two
[univac]$ wing compile echo.main.w                             
first
two
```

in may real program, i had to resort to JavaScript to generate output without a newline and to exit with a non-zero return code. i do not see any way to have `args: Array<str>?` from the command line passed to a program, e.g.:

```
[univac]$ cat echo.main.w 
let main = (args: Array<str>?) => {
  if let args = args {
    log(args.at(0));
    log(args.at(1));
  }
};
[univac]$ wing compile echo.main.w     
[univac]$ wing compile echo.main.w -- one two
[univac]$ 
```

i assume that simply isn't possible right now. am i correct?

(of course i do realize i'm, subverting the overall purpose of Wing to create a simple CLI program. nonetheless...)

**Comment by staycoolcall911** [comment]
Hi @mortax, I believe we currently don't have a straight forward way to reference arbitrary command line arguments on Wing code. You're more than welcome to open an issue with your use case.

As a workaround, if env variables could help in your case, that is actually supported in Wing and you can reference env variables in the following way:
1. Import util library by adding `bring util;` at the head of your `main.w`.
2. `util.env()` or `util.tryEnv()` will give you the value of the env variable ([docs](https://www.winglang.io/docs/standard-library/util/api-reference#@winglang/sdk.util.Util.tryEnv)).

**Comment by mortax** [comment]
thanks @staycoolcall911. using environment variables is cleaner that my current method so i'll give that a try. i'll open an issue as soon as i can.

**Reply by staycoolcall911** [reply]
Thank you, appreciated. Please let us know how it goes

**Reply by mortax** [reply]
shoot: i never replied. i used something like this:

```
let var args = MutArray<str>[];

if let env = Json.tryParse(util.tryEnv("WING_ARGS")) {
  for v in Json.values(env) {
    args.push(v.asStr());
  }
}
```

which worked very well. thanks.
